autoloader.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+/**
+ * Autoload the classes used by the AMP plugin.
+ *
+ * Class AMP_Autoloader
+ */
+class AMP_Autoloader {
+
+	/**
+	 * Map of Classname to relative filepath sans extension.
+	 *
+	 * @example Format:
+	 *
+	 *  array(
+	 *      'Class_Name1' =>  'subdir-of-includes/filename1',
+	 *      'Class_Name2' =>  '2nd-subdir-of-includes/filename2',
+	 *      'Class_Name3' =>  '!/full/path/to/filename3.php',
+	 *  );
+	 *
+	 * @var string[]
+	 */
+	private $_classmap;
+
+	/**
+	 * Provide access to the singlton instance
+	 *
+	 * @var AMP_Autoloader
+	 */
+	private static $_instance;
+
+	/**
+	 * Perform the autoload.
+	 *
+	 * Run as few lines of code as possible.
+	 *
+	 * If $class_name has first char as '!' strip it and assume a full path w/extension.
+	 * Otherwise assume subdir or /includes/ w/o extension.
+	 *
+	 * @param string $class_name
+	 */
+	function _autoload( $class_name ) {
+		if ( ! isset( $this->_classmap[ $class_name ] ) ) {
+			return;
+		}
+		$filepath = $this->_classmap[ $class_name ];
+		require( '!' !== $filepath[ 0 ] ? __DIR__ . "/includes/{$filepath}.php" : substr( $filepath, 1 ) );
+	}
+
+	/**
+	 * Trigger creation of AMP_Autoloader singleton instance and return it.
+	 *
+	 * @return AMP_Autoloader
+	 */
+	public static function instance() {
+		if ( ! isset( static::$_instance ) ) {
+			static::$_instance = new static();
+		}
+		return static::$_instance;
+	}
+
+	/**
+	 * Registers to autoloader to PHP.
+	 *
+	 * Called at the end of this file; calling a second time has no effect.
+	 */
+	public static function register() {
+		static $registered = false;
+		if ( ! $registered ) {
+			spl_autoload_register( [ static::instance(), '_autoload' ] );
```

Comment:
Syntax error in PHP 5.2

Reply:
5.2?  :-o
Fixed.

Reply:
Yeah. Currently the plugin supports PHP 5.2: https://github.com/Automattic/amp-wp/blob/eeb502b28807c1be80e59937fab6112e406b204b/.travis.yml#L45-L47

Reply:
Noted.